scio-elasticsearch/es8/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,791 @@
+        // assign this element to a random shard
+        final long shard = ThreadLocalRandom.current().nextLong(numOfShard);
+        c.output(KV.of(shard, c.element()));
+      }
+    }
+
+    private static class ElasticsearchWriter<T> extends DoFn<T, Void> {
+
+      private List<BulkOperation> chunk;
+      private long chunkBytes;
+      private final ClientSupplier clientSupplier;
+      private final SerializableFunction<T, Iterable<BulkOperation>> toBulkOperations;
+      private final ThrowingConsumer<BulkExecutionException> error;
+      private final int maxBulkRequestSize;
+      private final long maxBulkRequestBytes;
+      private final int maxRetries;
+      private final Duration retryPause;
+
+      private ProcessFunction<BulkRequest, BulkResponse> requestFn;
+      private ProcessFunction<BulkRequest, BulkResponse> retryFn;
+
+      public ElasticsearchWriter(
+          HttpHost[] nodes,
+          int maxBulkRequestSize,
+          long maxBulkRequestBytes,
+          SerializableFunction<T, Iterable<BulkOperation>> toBulkOperations,
+          ThrowingConsumer<BulkExecutionException> error,
+          int maxRetries,
+          Duration retryPause,
+          UsernamePasswordCredentials credentials,
+          JsonpMapperFactory mapperFactory) {
+        this.maxBulkRequestSize = maxBulkRequestSize;
+        this.maxBulkRequestBytes = maxBulkRequestBytes;
+        this.clientSupplier = new ClientSupplier(nodes, credentials, mapperFactory);
+        this.toBulkOperations = toBulkOperations;
+        this.error = error;
+        this.maxRetries = maxRetries;
+        this.retryPause = retryPause;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        checkArgument(
+            this.clientSupplier.get().ping().value(), "Elasticsearch client not reachable");
+
+        final FluentBackoff backoffConfig =
+            FluentBackoff.DEFAULT
+                .withMaxRetries(this.maxRetries)
+                .withInitialBackoff(this.retryPause);
+        this.requestFn = request(clientSupplier, error);
+        this.retryFn = retry(requestFn, backoffConfig);
+      }
+
+      @Teardown
+      public void teardown() throws Exception {
+        this.clientSupplier.getTransport().close();
+      }
+
+      @StartBundle
+      public void startBundle(StartBundleContext c) {
+        chunk = new ArrayList<>();
+        chunkBytes = 0;
+      }
+
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        flush();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        final T t = c.element();
+        final Iterable<BulkOperation> operations = toBulkOperations.apply(t);
+        for (BulkOperation operation : operations) {
+          long bytes = operationSize(clientSupplier, operation);
+          if ((chunk.size() + 1) > maxBulkRequestSize
+              || (chunkBytes + bytes) > maxBulkRequestBytes) {
```

Comment:
I'm wondering if we really need the `maxBulkRequestSize` option -- offering just the `maxBulkRequestBytes` functionality might be good enough for users. wdyt?

Reply:
Well, knowing the docs it's the best scenario but there are use cases where the nature of the payload is somewhat dynamic and vary greatly on size. That's why this combo `maxBulkRequestSize` + `maxBulkRequestBytes` was introduced. That said, not sure about ES8 but previous versions couldn't cope with this and it was really hard to keep/tune the ES clusters to have predictable indexing speeds and be stable.

Reply:
Will put that back then. Agreed that the doc example mention indexing only.
We can have cases where things are mixed and instead of being aggressive on the `maxBulkRequestSize` to ensure the request never gets too big, we should leave `maxBulkRequestBytes`

We discussed the settings naming with @clairemcginty because we think that `maxBulkRequestSize` vs `maxBulkRequestBytes` is confusing. `size` and `bytes` are often used for the same thing. I would suggest to rename `maxBulkRequestSize` option to `maxBulkRequestOperations` WDYT ?

Reply:
no opposition in renaming it; either that or perhaps `maxBulkRequestSizeBytes`